pymc/sampling.py

```diff
@@ -1534,6 +1542,166 @@ def stop_tuning(step):
+        as potential random variables.
+    givens_dict : Optional[Dict[aesara.graph.basic.Variable, Any]]
+        A dictionary that maps tensor variables to the values that should be used to replace them
+        in the compiled function. The types of the key and value should match or an error will be
+        raised during compilation.
+    """
+    if givens_dict is None:
+        givens_dict = {}
+
+    if basic_rvs is None:
+        basic_rvs = []
+
+    def is_a_random_variable(node):
+        return node in basic_rvs or (
+            isinstance(node, Variable)
+            and node.owner is not None
+            and isinstance(node.owner.op, RandomVariable)
+        )
+
+    def distribution_not_dist_parameters(node):
+        return [
+            rv_param
+            for rv_param in node.owner.inputs
+            if (
+                not (
+                    isinstance(  # Random state and generator shared variables are always volatile but they don't count as distribution parameters
+                        rv_param, (RandomStateSharedVariable, RandomGeneratorSharedVariable)
+                    )
+                    or (
+                        isinstance(
+                            getattr(getattr(rv_param, "owner", None), "op", None), RandomVariable
+                        )
+                        and rv_param not in basic_rvs
+                    )  # Mixture distributions have RandomVariable parameters that are not measurable (not in model.basic_RVs). These shouldn't propagate their volatility
+                )
+            )
+        ]
+
+    # We need a function graph to walk the clients and propagate the volatile property
+    fg = FunctionGraph(outputs=outputs, clone=False)
+
+    # Walk the graph from inputs to outputs and tag the volatile variables
+    nodes: List[Variable] = general_toposort(
+        fg.outputs, deps=lambda x: x.owner.inputs if x.owner else []
+    )
+    volatile_nodes: Set[Any] = set()
+    for node in nodes:
+        if isinstance(node, SharedVariable) or node in fg.outputs or node in givens_dict:
+            volatile_nodes.add(node)
+        is_volatile = node in volatile_nodes
+
+        # Propagate the volatile property to the node's clients
+        if is_volatile:
+            clients = [out for c, _ in fg.clients[node] if c != "output" for out in c.outputs]  # type: ignore
+            for client in clients:
+                if is_a_random_variable(client) and any(
+                    rv_param in volatile_nodes
+                    for rv_param in distribution_not_dist_parameters(client)
+                ):
+                    # Random variables that have volatile parameters are also volatile
+                    volatile_nodes.add(client)
+                elif not (client in vars_in_trace and is_a_random_variable(client)):
+                    # Other variables that aren't random variables in the trace are volatile
+                    volatile_nodes.add(client)
```

Comment:
@lucianopaz what about this? It's very similar but it is more strict about the sources of `volatility`, and after that any node that has a volatile input is considered volatile, regardless of what it is.

The reason I went to look for an alternative is the case where `dist`s depend on other standard RVs. With this code I think the following model would work as expected:

```python
with pm.Model() as m:
  latent_mu = pm.Normal("latent_mu")
  base_dist = pm.Normal.dist(latent_mu, 1)
  censored = pm.Censored("censored", base_dist, lower=-1, upper=1)
```
That is, if `latent_mu` is in the trace, `censored` will not be resampled by default, but if it is missing, it will, as `base_dist` will then be a volatile input. Do we have a test for this case?

```suggestion
    # We need a function graph to walk the clients and propagate the volatile property
    fg = FunctionGraph(outputs=outputs, clone=False)

    # Walk the graph from inputs to outputs and tag the volatile variables
    nodes: List[Variable] = general_toposort(
        fg.outputs, deps=lambda x: x.owner.inputs if x.owner else []
    )
    volatile_nodes: Set[Any] = set()
    for node in nodes:
        if (
            # SharedVariables, except RandomState/Generators
            (isinstance(node, SharedVariable) and not isinstance(node, (RandomStateSharedVariable, RandomGeneratorSharedVariable)))
            # Basic RVs that are not in the trace
            or (node.owner and isinstance(node.owner.op, RandomVariable) and node not in vars_in_trace and node in basic_rvs)
            # Variables that have any volatile input
            or (node.owner and any(inp in volatile_nodes for inp in node.owner.inputs))
            or node in fg.outputs
            or node in givens_dict
        ):
            volatile_nodes.add(node)

```